pkg/setting/settings.go

```diff
@@ -0,0 +1,140 @@
+package setting
+
+import (
+	"errors"
+	"gopkg.in/ini.v1"
+	"strings"
+	"time"
+)
+
+var (
+	ErrOperationNotPermitted = errors.New("operation not permitted")
+)
+
+type ValidationError struct {
+	Errors []error
+}
+
+func (v ValidationError) Error() string {
+	builder := strings.Builder{}
+
+	for i, e := range v.Errors {
+		builder.WriteString(e.Error())
+		if i != len(v.Errors)-1 {
+			builder.WriteString(", ")
+		}
+	}
+
+	return builder.String()
+}
+
+// Provider is a settings provider abstraction
+// with thread-safety and runtime updates.
+type Provider interface {
+	// Update
+	Update(updates SettingsBag, removals SettingsRemovals) error
+	// KeyValue returns a key-value abstraction
+	// for the given pair of section and key.
+	KeyValue(section, key string) KeyValue
+	// Section returns a settings section
+	// abstraction for the given section name.
+	Section(section string) Section
+	// RegisterReloadHandler registers a handler for validation and reload
+	// of configuration updates tied to a specific section
+	RegisterReloadHandler(section string, handler ReloadHandler)
+}
+
+// Section is a settings section copy
+// with all of its pairs of keys-values.
+type Section interface {
+	// KeyValue returns a key-value
+	// abstraction for the given key.
+	KeyValue(key string) KeyValue
+}
+
+// KeyValue represents a settings key-value
+// for a given pair of section and key.
+type KeyValue interface {
+	// Key returns pair's key.
+	Key() string
+	// Value returns pair's value.
+	Value() string
+
+	// MustString returns the value's string representation
+	// If empty, then it returns the given default.
+	MustString(defaultVal string) string
+	// MustBool returns the value's boolean representation
+	// Otherwise returns the given default.
+	MustBool(defaultVal bool) bool
+	// MustDuration returns the value's time.Duration
+	// representation. Otherwise returns the given default.
+	MustDuration(defaultVal time.Duration) time.Duration
+}
+
+// ReloadHandler makes it possible
```

Comment:
```suggestion
// ReloadHandler makes it possible to reload services after configuration changes
```
Not sure about what you wanted to write here.

Reply:
I actually don't remember, maybe because it was Leo 😅 
Anyway, I'll write some explanation. Good catch, thanks! 🙏🏻